File: python/mfd_solver.py
import numpy as np

'''
Class Solver: solving equation:
∂U/∂t - div (D · grad U) = f(x, y, z, t) on [0; 1]³ and [0, T]

Using Diag tensor D: 
	    | d_x  0   0  |
	D = |  0  d_y  0  |
	    |  0   0  d_z |
where d_x = 0.25, d_y = 0.15, d_z = 0.1

Also we have 
g(x, y, z) = 0 - boundary condition
f(x, y, z) = (d_x+d_y+d_z)·π²·sin(πx)sin(πy)sin(πz) - right side
discretization:
n,m,k - number of nodes for x,y,z directions
our equation have analytic solve:
U_analityc = sin(πx)sin(πy)sin(πz)·(1 - exp(-(d_x+d_y+d_z)·π²·t))
Steps:
hx = 1/(Nx-1), hy = 1/(Ny-1), hz = 1/(Nz-1).

also dt is time step has condition:
dt < 0.5 / { d_x / (hx)² + d_y / (hy)² + d_z / (hz)² }

Solving:
Let us introduce discrete operators of secondary glass derivatives:
	Lx_ijk U^n = ( U_(i-1)jk^n - 2·U_ijk^n + U_(i+1)jk^n )/(hx)²
	Ly_ijk U^n = ( U_i(j-1)k^n - 2·U_ijk^n + U_i(j+1)k^n )/(hy)²
	Lz_ijk U^n = ( U_ij(k-1)^n - 2·U_ijk^n + U_ij(k+1)^n )/(hz)²
Also let: 
    f_ijk^n = f(i·hx, j·hy, k·hz, ndt) и g_ijk = g(i·hx, j·hy, k·hz)
    
And, finally our numerical scheme:
U_ijk^(n+1) = U_ijk^n + Δt·(f_ijk^n + d_x·Lx_ijk U^n + d_y·Ly_ijk U^n + d_z·Lz_ijk U^n), 
if 1 ⩽ i ⩽ Nx-2, 1 ⩽ j ⩽ Ny-2, 1 ⩽ k ⩽ Nz-2 (not boundary)
	U_ijk^(n+1) = g_ijk, if (i%(Nx-1)) · (j%(Ny-1)) · (k%(Nz-1)) = 0, (boundary)


'''


class Solver:
    D = [0.25, 0.15, 0.1]

    def __init__(self, n, m, k, t):
        self.n = n
        self.m = m
        self.k = k
        self.t = t
        self.dt = 1 / t
        self.u_actual = np.zeros((n, m, k))
        self.u_next = np.zeros((n, m, k))
        self.hx = 1 / (n - 1)
        self.hy = 1 / (m - 1)
        self.hz = 1 / (k - 1)

    def check_boundary(self, i, j, k):
        if i == 0 or i == self.n \
                or j == 0 or i == self.m \
                or k == 0 or k == self.k:
            return True
        else:
            return False

    def fill_boundaries(self):
        bi = [0, self.n - 1]
        bj = [0, self.m - 1]
        bk = [0, self.k - 1]
        for i in bi:
            for j in bj:
                for k in bk:
                    self.u_actual[i, j, k] = 0

    def Lx(self, i, j, k):
        return (self.u_actual[i - 1][j][k] - 2 * self.u_actual[i][j][k] + self.u_actual[i + 1][j][k]) \
            / (self.hx ** 2)

    def Ly(self, i, j, k):
        return (self.u_actual[i][j - 1][k] - 2 * self.u_actual[i][j][k] + self.u_actual[i][j + 1][k]) \
            / (self.hy ** 2)

    def Lz(self, i, j, k):
        return (self.u_actual[i][j][k - 1] - 2 * self.u_actual[i][j][k] + self.u_actual[i][j][k + 1]) \
            / (self.hz ** 2)

    # right side
    def f(self, x, y, z):
        return ((self.D[0] + self.D[1] + self.D[2])
                * (np.pi ** 2) * np.sin(np.pi * x) * np.sin(np.pi * y) * np.sin(np.pi * z))

    # calculating u on next t step
    def get_next_u(self):
        for i in range(1, self.n - 1):
            for j in range(1, self.m - 1):
                for k in range(1, self.k - 1):
                    self.fill_boundaries()
                    # print('NOT boundary i={}, j={}, k={}'.format(i, j, k))
                    # print('F: {}'.format(self.f(i * self.hx, j * self.hy, k * self.hz)))
                    self.u_next[i, j, k] = self.u_actual[i, j, k] + \
                                           self.dt * (self.f(i * self.hx, j * self.hy, k * self.hz)
                                                      + self.D[0] * self.Lx(i, j, k)
                                                      + self.D[1] * self.Ly(i, j, k)
                                                      + self.D[2] * self.Lz(i, j, k))
        # print(self.u_next)

    # putting u on next t step in u_actual
    # REMEMBER: actual u in u_actual
    def calculate_next_u(self):
        self.get_next_u()
        # tmp = self.u_next
        # self.u_next = self.u_actual
        self.u_actual = self.u_next

    # solver function
    # call it to get answer
    def solve(self):
        for i in range(1, self.t):
            self.calculate_next_u()
        return self.u_actual

    def get_analytic_u(self, x, y, z):
        return (np.sin(np.pi * x)
                * np.sin(np.pi * y)
                * np.sin(np.pi * z)
                * (1 - np.exp(-((self.D[0] + self.D[1] + self.D[2]) * np.pi ** 2 * 1))))

    def get_analytic_solve(self):
        u = np.zeros((self.n, self.m, self.k))
        for i in range(1, self.n - 1):
            for j in range(1, self.m - 1):
                for k in range(1, self.k - 1):
                    self.fill_boundaries()
                    # print('NOT boundary i={}, j={}, k={}'.format(i, j, k))
                    # print('F: {}'.format(self.f(i * self.hx, j * self.hy, k * self.hz)))
                    u[i, j, k] = self.get_analytic_u(i * self.hx, j * self.hy, k * self.hz)
        return u



    def print(self):
        print('n = {}, m = {}, k = {}'.format(self.n, self.m, self.k))
        print('hx = {}, hy = {}, hz = {}'.format(self.hx, self.hy, self.hz))
        print('U: ')
        # print(self.u_actual)
        print('dt=',self.dt,' < ',0.5/(self.D[0]/(self.hx**2)+self.D[1]/(self.hy**2)+self.D[2]/(self.hz**2)))



def max_mistake(a,b):
    mist = 0
    our = a
    anal = b
    for i in range(len(a)):
        for j in range(len(a)):
            for k in range(len(a)):
                mist = max(np.abs(our[i, j, k] - anal[i, j, k]), mist)
    return mist

File: CPP/mfd_solver.cpp
#include "mfd_solver.h"
#include <cmath>

Solver::Solver(int N, int M, int K, int T) {
    this->n = N;
    this->m = M;
    this->kk = K;
    this->t = T;
    this->hx = 1. / (n - 1) * 1.0;
    this->hy = 1. / (m - 1) * 1.0;
    this->hz = 1. / (kk - 1) * 1.0;
    this->dt = 1. / T * 1.0;

    for (int i = 0; i < this->n; ++i) {
        for (int j = 0; j < this->m; ++j) {
            for (int k = 0; k < this->kk; ++k) {
                this->u_actual->push_back(0.);
                this->u_next->push_back(0.);
            }
        }
    }
}

double const &Solver::operator()(int i, int j, int k) {
    return this->u_actual->at(i * this->n * this->m + j * this->kk + k);
}

bool Solver::check_boundary(int I, int J, int K) {
    if ((I == 0) || (I == this->n) || (J == 0) || (J == this->m) ||
        (K == 0) || (K == this->kk))
        return true;
    else
        return false;
}


void Solver::check_t() {
    std::cout << "dt = " << this->dt << " < " << 0.5 /
                                                 (this->D[0] / (pow(this->hx, 2)) + this->D[1] / (pow(this->hy, 2)) +
                                                  this->D[2] / (pow(this->hz, 2))) << std::endl;
    std::string answer;
    std::cout << "Do you want to continue: (y/n):";
    std::cin >> answer;

    if (answer == "y") {

    } else if (answer == "n") {
        std::cout << "enter new t:";
        std::cin >> this->t;
        this->dt = 1. / this->t;
        this->check_t();
    } else {
        std::cerr << "wrong input" << std::endl;
        exit(1);
    }
}

std::vector<double> *Solver::solve() {
    this->check_t();
    for (int i = 1; i <= this->t; i++) {
//        print();
//        std::cout<<i<<std::endl;
        calculate_next_u();
    }
    return this->u_actual;
}

void Solver::fill_boundary_x() {
    for (int j = 0; j < this->m; ++j) {
        for (int k = 0; k < this->kk; ++k) {
            set_value(0, j, k, 0.);
            set_value(this->n - 1, j, k, 0.);
        }
    }
}

void Solver::fill_boundary_y() {
    for (int i = 0; i < this->n; ++i) {
        for (int k = 0; k < this->kk; ++k) {
            set_value(i, 0, k, 0.);
            set_value(i, this->m - 1, k, 0.);
        }
    }
}


void Solver::fill_boundary_z() {
    for (int i = 0; i < this->n; ++i) {
        for (int j = 0; j < this->m; ++j) {
            set_value(i, j, 0, 0.);
            set_value(i, j, this->kk - 1, 0.);
        }
    }
}


void Solver::fill_boundaries() {
    fill_boundary_x();
    fill_boundary_y();
    fill_boundary_z();
}


double Solver::get_analytic_u(int i, int j, int k) const {
    double x = i * this->hx;
    double y = j * this->hy;
    double z = k * this->hz;
    return sin(M_PI * x) * sin(M_PI * y) * sin(M_PI * z) *
           (1 - exp(-((this->D[0] + this->D[1] + this->D[2]) * pow(M_PI, 2) * 1)));
}


void Solver::set_value(int i, int j, int k, const double &value) {
    this->u_actual->at(i * this->n * this->m + j * this->kk + k) = value;
}

void Solver::set_value_next(int i, int j, int k, const double &value) {
    this->u_next->at(i * this->n * this->m + j * this->kk + k) = value;
}

double Solver::get_u(int i, int j, int k) const {
    return this->u_actual->at(i * this->n * this->m + j * this->kk + k);
}


inline double Solver::Lx(int i, int j, int k) {
    return (get_u(i - 1, j, k) - 2 * get_u(i, j, k) + get_u(i + 1, j, k))
           / (this->hx * this->hx);
}

inline double Solver::Ly(int i, int j, int k) {
    return (get_u(i, j - 1, k) - 2 * get_u(i, j, k) + get_u(i, j + 1, k))
           / (this->hy * this->hy);
}

inline double Solver::Lz(int i, int j, int k) {
    return (get_u(i, j, k - 1) - 2 * get_u(i, j, k) + get_u(i, j, k + 1))
           / (this->hz * this->hz);
}

inline double Solver::f(int i, int j, int k) {
    double x = i * this->hx;
    double y = j * this->hy;
    double z = k * this->hz;
    return ((this->D[0] + this->D[1] + this->D[2])
            * (M_PI * M_PI) * sin(M_PI * x) * sin(M_PI * y)
            * sin(M_PI * z));
}

void Solver::swap() {
    u_actual->swap(*u_next);
}

void Solver::get_next_u() {
    fill_boundaries();
    double u;
    for (int i = 1; i < this->n - 1; ++i) {
        for (int j = 1; j < this->m - 1; ++j) {
            for (int k = 1; k < this->kk - 1; ++k) {
                u = get_u(i, j, k) +
                    this->dt * (f(i, j, k)
                          + D[0] * Lx(i, j, k)
                          + D[1] * Ly(i, j, k)
                          + D[2] * Lz(i, j, k));
//                std::cout << "dt: " << this->dt << " get_u: " << get_u(i, j, k) << " " << " f(i, j, k): " << f(i, j, k) <<
//                          " D[0] * Lx(i, j, k): " << +D[0] * Lx(i, j, k) << " D[1] * Ly(i, j, k): "
//                          << D[1] * Ly(i, j, k)
//                          << " D[2] * Lz(i, j, k): " << D[2] * Lz(i, j, k) << " u: "
//                          << u << std::endl;
                set_value_next(i, j, k, u);
            }
        }
    }
}

inline void Solver::calculate_next_u() {
    get_next_u();
    swap();
}

double Solver::mistake() const {
    double mist = 0.;
    for (int i = 1; i < this->n - 1; ++i) {
        for (int j = 1; j < this->m - 1; ++j) {
            for (int k = 1; k < this->kk - 1; ++k) {
//                std::cout<<std::abs(this->get_u(i, j, k) - this->get_analytic_u(i, j, k))<< std::endl;
                if (std::abs(this->get_u(i, j, k) - this->get_analytic_u(i, j, k)) > mist) {
                    mist = std::abs(get_u(i, j, k) - get_analytic_u(i, j, k));
                }
            }
        }
    }
    return mist;
}


